homeassistant/auth/mfa_modules/totp.py

```diff
@@ -0,0 +1,209 @@
+"""Time-based One Time Password auth module."""
+import logging
+from typing import Any, Dict, Optional, Tuple  # noqa: F401
+
+import voluptuous as vol
+
+from homeassistant.auth.models import User
+from homeassistant.core import HomeAssistant
+from homeassistant.loader import bind_hass
+
+from . import MultiFactorAuthModule, MULTI_FACTOR_AUTH_MODULES, \
+    MULTI_FACTOR_AUTH_MODULE_SCHEMA, SetupFlow
+
+REQUIREMENTS = ['pyotp==2.2.6', 'PyQRCode==1.2.1', 'pypng==0.0.18']
+
+CONFIG_SCHEMA = MULTI_FACTOR_AUTH_MODULE_SCHEMA.extend({
+}, extra=vol.PREVENT_EXTRA)
+
+STORAGE_VERSION = 1
+STORAGE_KEY = 'auth_module.totp'
+STORAGE_USERS = 'users'
+STORAGE_USER_ID = 'user_id'
+STORAGE_OTA_SECRET = 'ota_secret'
+
+INPUT_FIELD_CODE = 'code'
+
+DUMMY_SECRET = 'FPPTH34D4E3MI2HG'
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _generate_qr_code(data: str) -> str:
+    """Generate a base64 PNG string represent QR Code image of data."""
+    import pyqrcode
+
+    qr_code = pyqrcode.create(data)
+    return str(qr_code.png_as_base64_str(scale=4))
+
+
+@bind_hass
+async def _async_generate_secret(hass: HomeAssistant, username: str) \
+        -> Tuple[str, str, str]:
+    """Generate a secret, url, and QR code."""
+    def generate_secret_helper(username: str) -> Tuple[str, str, str]:
+        """Generate a secret, url, and QR code."""
+        import pyotp
+
+        ota_secret = pyotp.random_base32()
+        url = pyotp.totp.TOTP(ota_secret).provisioning_uri(
+            username, issuer_name="Home Assistant")
+        image = _generate_qr_code(url)
+        return ota_secret, url, image
+
+    return await hass.async_add_executor_job(
```

Comment:
If the whole method is about running a sync method, just make the method sync on the global scope and call with `async_add_executor_job` directly.